### PR TITLE
fix: [PL-39412]: Add try/catch block on get cache from redis

### PR DESCRIPTION
**Pull request by jennyjames21**
## Describe your changes
Summary:
-  Issue :  Kryo exception on fetching delegate group cache from Redis due to new field in the DelegateGroup. Currently we do not have backward compatibility support for the new fields in RedissonKryoCodec   

Fix: Add try/catch block for any exceptions so that it will force to load from DB

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49142)

### CODE REVIEW

**Review by xingchi-jin — COMMENTED** [review]
Thanks for the quick fix

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java [hunk 1]
```diff
@@ -343,19 +343,33 @@ private long getDelegateTaskCount(String accountId, DelegateTaskRank rank, boole
...
 
   private Delegate getDelegateFromRedisCache(String delegateId, boolean forceRefresh) {
-    if (delegateRedisCache.get(delegateId) == null || forceRefresh) {
-      Delegate delegate = persistence.createQuery(Delegate.class).filter(DelegateKeys.uuid, delegateId).get();
-      if (delegate == null) {
-        log.warn("Unable to find delegate {} in DB.", delegateId);
-        return null;
-      }
-      delegateRedisCache.put(delegateId, delegate);
+    Delegate delegateFromCache = null;
```
[the remark this answers is not in the document]
Comment (jennyjames21) [reply 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java]: updated

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java [hunk 2]
```diff
@@ -343,19 +343,36 @@ private long getDelegateTaskCount(String accountId, DelegateTaskRank rank, boole
   }
 
   private Delegate getDelegateFromRedisCache(String delegateId, boolean forceRefresh) {
-    if (delegateRedisCache.get(delegateId) == null || forceRefresh) {
-      Delegate delegate = persistence.createQuery(Delegate.class).filter(DelegateKeys.uuid, delegateId).get();
-      if (delegate == null) {
-        log.warn("Unable to find delegate {} in DB.", delegateId);
-        return null;
+    if (!forceRefresh) {
```
Comment (raghuAtWings) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java:346]: this is pretty complicated, can we not always read from db if there is an exception. I would recommend to rewrite this the following way

if(forceRefresh) {
   //read from db,  update the cache and return 
}

try {
    //read from cache and return
} catch (Exception) {
   //read from db,  update the cache and return 
}

**Review by raghuAtWings — CHANGES_REQUESTED** [review]

**Comment by jennyjames21** [comment]
trigger uts

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java [hunk 3]
```diff
@@ -343,7 +343,20 @@ private long getDelegateTaskCount(String accountId, DelegateTaskRank rank, boole
...
+        return null;
+      }
+      delegateRedisCache.put(delegateId, delegate);
+      return delegateRedisCache.get(delegateId);
+    }
+
+    try {
+      delegateRedisCache.get(delegateId);
+    } catch (Exception ex) {
+      log.error("Exception occurred while loading Delegate from redis cache, loading from DB", ex);
```
Comment (raghuAtWings) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java:359]: should not be error, info should be good enough.

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java [hunk 4]
```diff
@@ -355,7 +368,14 @@ private Delegate getDelegateFromRedisCache(String delegateId, boolean forceRefre
   }
 
   private DelegateGroup getDelegateGroupRedisCache(String accountId, String delegateGroupId) {
-    if (delegateGroupRedisCache.get(delegateGroupId) == null) {
+    DelegateGroup delegateGroupFromCache = null;
+    try {
+      delegateGroupFromCache = delegateGroupRedisCache.get(delegateGroupId);
+    } catch (Exception ex) {
+      log.error("Exception occurred while loading Delegate Group from redis cache", ex);
```
Comment (raghuAtWings) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java:375]: same here.

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java [hunk 5]
```diff
@@ -366,7 +386,14 @@ private DelegateGroup getDelegateGroupRedisCache(String accountId, String delega
   }
 
   private List<Delegate> getDelegatesForGroupRedisCache(String accountId, String delegateGroupId) {
-    if (delegatesFromGroupRedisCache.get(delegateGroupId) == null) {
+    List<Delegate> delegateListFromGroupCache = null;
+    try {
+      delegateListFromGroupCache = delegatesFromGroupRedisCache.get(delegateGroupId);
+    } catch (Exception ex) {
+      log.error("Exception occurred while loading Delegate list from redis cache", ex);
```
Comment (raghuAtWings) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java:393]: same here

**Review by raghuAtWings — CHANGES_REQUESTED** [review]

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java (hunk 5 above)
Comment (jennyjames21) [reply 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java]: updated

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java (hunk 4 above)
Comment (jennyjames21) [reply 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java]: updated

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java (hunk 3 above)
Comment (jennyjames21) [reply 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java]: updated

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java (hunk 2 above)
Comment (jennyjames21) [reply 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java]: updated

**Review by raghuAtWings — APPROVED** [review]

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger messagecheck

**Pull request merged**